File: LinkList/doublelist.c
#include <stdio.h>
#include <stdlib.h>

// 双链循环链表

// 节点
struct Node
{
    int val;
    struct Node *prior; // 指向上一个节点的地址
    struct Node *next;  // 指向下一个节点的地址
};

// 返回一个节点
typedef struct Node newnode;
// 返回一个节点的地址
typedef struct Node *nodeAddress;

// 空链表
nodeAddress InitList(int val, nodeAddress *head)
{
    // 创建一个新的节点
    nodeAddress newnode = (nodeAddress)malloc(sizeof(struct Node));
    (*head)->next = newnode;
    newnode->next = newnode;
    newnode->val = val;
    // newnode -> next = NULL;
    newnode->prior = newnode;
    return newnode;
}

// 从指定位置处插入节点
int InsertListBylocation(int val, nodeAddress head, int index)
{
    // 新增节点
    nodeAddress newnode = (nodeAddress)malloc(sizeof(struct Node));
    newnode->val = val;
    newnode->next = NULL;
    newnode->prior = NULL;

    // 判断终止条件
    if (head == NULL || index < 0)
    {
        return 0;
    }

    // 插入表头
    if (index == 1)
    {
       
        newnode->next = head;
        head->prior = newnode;
    }
    // 创建游标
    nodeAddress current = head;
    for (int i = 0; i < index - 1; i++)
    {
        current = current->next;
    }
    // 判断节点位置
    if (current->next == NULL)
    {
        // 说明是尾节点
        current->next = newnode;
        newnode->prior = current;
    }
    else
    {
        // 说明是中间节点
        newnode->prior = current;
        newnode->next = current->next;
        current->next->prior = newnode;
        current->next = newnode;
    }

    return 1;
}

// 遍历输出
int TraverseList(nodeAddress head)
{
    if (head == NULL)
    {
        return 0;
    }
    nodeAddress current = head;
    while (current)
    {
        printf("%d\n", current->val);
        current = current->next;
    }
    return 0;
}


int main()
{
    struct Node *L;
    L = InitList(2, &L);
    // AddListTaril(3,L);
    InsertListBylocation(3, L, 1);
    InsertListBylocation(3, L, 2);
    TraverseList(L);
    return 0;
}

File: LinkList/doublelist.py

# 创建节点


class Node(object):
    def __init__(self,data=None,next=None):
        self.data = data
        self.next = None
        self.pre = None
    
    def __repr__(self):
        return str(self.data)

# 创建链表
class DoubleList():
    def __init__(self):
        self.head = None
    # 判断是否为空
    def isEmpty(self):
        return self.head is None
    # 初始化双链表
    def InitList(self,value:int):
        self.head = Node(value)
        self.head.next = self.head
        self.head.pre = self.head

    
    # 追加节点 尾插法
    def AddList(self,value:int):
        # 创建节点
        newnode = Node(value)
        # 关联节点
        current = self.head
        # 由于双向链表的特殊性，尾插法的尾节点=self.head.pre
        while current.next != self.head:
            current = current.next
        current.next = newnode
        newnode.pre = current
        newnode.next = self.head
        self.head.pre = newnode

    # 获取链表长度
    def __len__(self):
        count = 1
        current = self.head
        while current.next != self.head:
            count +=1
            current = current.next
        print(count)
    # 循环输出节点
    def ConsoleElement(self):
        current = self.head
        while current.next != self.head:
            print(current.data)
            current = current.next
        print(current.data)


    # 删除节点
    def DelList(self,index:int):
        current = self.head
        if index < 0 or self.head == None:
            return 0
        while index:
            current = current.next
            index -=1
        
        tmpnode = current.next

        current.next = current.next.next
        current.next.pre = current
        tmpnode.next = None
        tmpnode.pre = None

    # 修改节点
    def getIndexList(self,index:int,value:int):
        current = self.head
        if index < 0 or self.head ==None:
            return 0
        while index :
            current = current.next
            index -=1
        current.data =  value

if __name__ == '__main__':
    dl = DoubleList()
    dl.InitList(20)
    dl.AddList(30)
    dl.AddList(30)
    dl.__len__()
    dl.DelList(1)
    dl.__len__()
    dl.ConsoleElement()
    dl.getIndexList(1,20)
    dl.ConsoleElement()

    